File: src/recursion/print_all_sub_squences2.cpp
#include <iostream>
#include <unordered_set>
using namespace std;

class Solution {
 private:
  void process(string s, int index, unordered_set<string>& set, string path) {
    if (index == s.size()) {
      set.insert(path);
      return;
    }

    string no = path;
    process(s, index + 1, set, no);
    string yes = path + s[index];
    process(s, index + 1, set, yes);
  };

 public:
  vector<string> subNoRepeat(string s) {
    string path = "";
    unordered_set<string> set;
    process(s, 0, set, path);
    vector<string> ans;
    for (string cur : set) ans.push_back(cur);
    return ans;
  }
};

int main() {
  Solution s;
  string test = "acccc";
  vector<string> ans = s.subNoRepeat(test);

  for (string str : ans) {
    cout << str << endl;
  }
}

File: src/recursion/print_all_sub_squences2.java
package recursion;

import java.util.ArrayList;
import java.util.HashSet;
import java.util.List;

public class print_all_sub_squences2 {
  public static List<String> subNoRepeat(String s) {
    char[] str = s.toCharArray();
    String path = "";
    HashSet<String> set = new HashSet<>();
    process(str, 0, set, path);
    List<String> ans = new ArrayList<>();
    for (String cur : set) {
      ans.add(cur);
    }
    return ans;
  }

  private static void process(char[] str, int index, HashSet<String> set, String path) {
    if (index == str.length) {
      set.add(path);
      return;
    }
    String no = path;
    process(str, index + 1, set, no);
    String yes = path + String.valueOf(str[index]);
    process(str, index + 1, set, yes);
  }

  public static void main(String[] args) {
    String test = "acccc";
    List<String> ans1 = subNoRepeat(test);

    for (String str : ans1) {
      System.out.println(str);
    }
    System.out.println("=================");
  }
}
